CRM/Contribute/BAO/ContributionRecur.php

```diff
@@ -940,15 +940,15 @@ public static function updateOnNewPayment($recurringContributionID, $paymentStat
 
     if ($paymentStatus == 'Completed'
       && CRM_Contribute_PseudoConstant::contributionStatus($existing['contribution_status_id'], 'name') == 'Pending') {
-      $params['contribution_status_id'] = 'In Progress';
+      $params['contribution_status_id'] = CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'In Progress');
```

Comment:
I'd say the tests are failing because contributionrecur has different statuses than contribution. But also, the fact that the tests were passing before means the api3 feature to use the name is still working. Is it possible you've disabled or are somehow missing the contribution_recur statuses in your db? If you go to admin - system settings - option groups do you have one for `contribution_recur_status` - does it have choices?

Reply:
@demeritcowboy huh, my db was missing the contribution_recur_status options. (I ran into another problem where it wouldn't show me any types for adding option values; maybe there is another missing option value, but added it directly to the db). I've added contribution_recur_status now.

@colemanw I tried to rebase my PR but didn't quite seem to manage to do it. I'd thought that I had done it, but apparently not. I have created a rebased commit on 5.70 at https://github.com/civicrm/civicrm-core/commit/61cec1b5fd2ff2b9bf42d621da3b2abfe5322444 but I don't know how to update this PR to point to that commit.

Reply:
Since it sounds like there is no code bug and this PR is about futureproofing now, it could just be against master.

Reply:
Hmm, maybe my assumption was incorrect and there's no bug here at all.

Reply:
Well, the stripe plugin is using v3 almost everywhere that I saw. It was definitely throwing a type error trying to put "In Progress" as the contribution_status_id, but my db was also missing that option value. I stopped looking after this, but I don't know where "In Progress" would get looked up as a contribution_recur_status option value. I guess it does somewhere. 

I do see that the BAO Contribution code does explicitly go through CRM_Core_PseudoConstant though; looks like ContributionRecur is older code that didn't also get updated.

Reply:
It happens at https://github.com/civicrm/civicrm-core/blob/3cab30bb5bf1be43f460b454ab453d5da5e3ac32/api/v3/utils.php#L2378-L2382.

Here's two options:
* Leave it as-is since the code is ok and the error seems related to missing config.
* Use getKey but using the right BAO (ContributionRecur), but it's unnecessary because of the above - api3 will convert by itself.

By the way using `$params['contribution_status_id:name'] = "In Progress"` doesn't seem to work with api3 in a quick test. It gets ignored and returns all the records.

Reply:
I see the lookup code, thanks.

I'll leave that up to the maintainers; it's not my codebase. The patch below does use the ContributionRecur BAO and it does work for me.